keps/sig-node/4369-allow-special-characters-environment-variable/README.md

```diff
@@ -0,0 +1,328 @@
+- [ ] (R) Production readiness review approved
+- [ ] "Implementation History" section is up-to-date for milestone
+- [ ] User-facing documentation has been created in [kubernetes/website], for publication to [kubernetes.io]
+- [ ] Supporting documentation—e.g., additional design documents, links to mailing list discussions/SIG meetings, relevant PRs/issues, release notes
+
+[kubernetes.io]: https://kubernetes.io/
+[kubernetes/enhancements]: https://git.k8s.io/enhancements
+[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
+[kubernetes/website]: https://git.k8s.io/website
+
+## Summary
+
+Allow all printable ASCII except '=', to be set as environment variables, the printable ASCII character range is serial number 33-126
+
+## Motivation
+
+Kubernetes should not restrict which environment variable names can be used, because it has no way of knowing what the application may need, and people can't always choose their own variable names, which may limit the adoption of Kubernetes.
+
+### Goals
+
+* Allows users to set all ASCII characters with serial numbers in the range of 33-126 except "=" as environment variables.
+
+### Non-Goals
+
+## Proposal
+
+* Implements relaxed validation at the top level validation method when validating API create requests, all printable ASCII characters can be verified, and their serial number range is 33-126
+* Allow users to set `Configmap` keys and secret keys outside the `C_IDENTIFIER` scope as environment variables using EnvFrom
+* Document rules for setting environment variables.
+
+### User Stories (Optional)
+
+#### Story 1
+
+I am a .NET Core development engineer, .Net Core applications are using ":" when working with application settings loaded from appsettings.json file. When running .net core app in containers typically overwrite this settings by specifying environmental variable.
+such as: 
+`"Logging": { "IncludeScopes": false, "LogLevel": { "Default": "Warning" } }`    
+override like this `-e Logging:LogLevel:Default=Debug`    
+
+### Risks and Mitigations
+
+Relaxed validation can break upgrade and rollback scenarios, but our use of feature gate to control whether it's enabled or not will make it a manageable risk, with the user having the autonomy to choose whether or not to enable it.
+
+## Design Details
+
+- A feature gate name `RelaxedEnvironmentVariableValidation` controlling the loosening of the envvar name validation, initially in alpha state and defaulting to false
+
+- Two sets of validation logic for envvar names:
+  
+  * Strict validation
+    * Strict validation follows the current design, which only allows envvar names passed the regular expression `[-._a-zA-Z][-._a-zA-Z0-9]*`.
+  
+  - Relaxed validation
+    * Relaxed verification allows all ASCII characters with numbers in the range 33-126 as envvar name, and its regular expression is `^[!-~]{1,}$`,  matches string that contains ASCII characters from the `!` to the `~` and is at least 1 in length.
```

Comment:
Maybe describe this (or implement it) as `unicode.IsPrint(r) && r < unicode.MaxASCII`

Reply:
We also need to filter out "space" and "=" from all printable ASCII characters, it will be easier to use regex.

Reply:
The reason I prefer `unicode.IsPrint(r) && r < unicode.MaxASCII` to a range or regex is that it's not reasonable to expect people to keep the ASCII range in their heads.

Reply:
all right, I agree with your opinion.